File: ui/src/lib/audit/backends/local-backend.ts
// assisted-by claude code claude-sonnet-4-6
/**
 * Local-disk audit log backend.
 *
 * Appends one JSON event per line (NDJSON) to:
 *   <path>/YYYY/MM/DD/<event_type>-<YYYYMMDD>-<uuid>.ndjson
 *
 * Files are opened in append mode. The UUID suffix is stable for the
 * lifetime of the process instance, keeping one file per (process × type × day).
 */

import fs from "fs/promises";
import path from "path";
import { randomUUID } from "crypto";
import type { AuditBackend } from "../backend";

function toIsoDate(ts: unknown): string {
  if (ts instanceof Date) return ts.toISOString();
  if (typeof ts === "string") return ts;
  return new Date().toISOString();
}

function zeroPad(n: number): string {
  return String(n).padStart(2, "0");
}

export class LocalBackend implements AuditBackend {
  private readonly root: string;
  private readonly fileUuid: string;

  constructor(rootPath: string) {
    this.root = rootPath;
    this.fileUuid = randomUUID().replace(/-/g, "").slice(0, 12);
  }

  write(event: Record<string, unknown>): void {
    void this._writeAsync(event);
  }

  private async _writeAsync(event: Record<string, unknown>): Promise<void> {
    try {
      const isoTs = toIsoDate(event["ts"]);
      const d = new Date(isoTs);
      const yyyy = String(d.getUTCFullYear());
      const mm = zeroPad(d.getUTCMonth() + 1);
      const dd = zeroPad(d.getUTCDate());
      const yyyymmdd = `${yyyy}${mm}${dd}`;

      const eventType = String(event["type"] ?? "audit");
      const dateDir = path.join(this.root, yyyy, mm, dd);
      await fs.mkdir(dateDir, { recursive: true });

      const filename = `${eventType}-${yyyymmdd}-${this.fileUuid}.ndjson`;
      const filepath = path.join(dateDir, filename);

      const serialized = JSON.stringify(event, (_k, v) =>
        v instanceof Date ? v.toISOString() : v
      );
      await fs.appendFile(filepath, serialized + "\n", "utf-8");
    } catch (err) {
      console.warn("[audit/local] Failed to write audit event:", err);
    }
  }
}

File: ai_platform_engineering/utils/audit_backends/local_backend.py
# Copyright 2025 CNOE Contributors
# SPDX-License-Identifier: Apache-2.0
"""Local-disk audit log backend.

Writes one JSON event per line (NDJSON) to::

    <path>/YYYY/MM/DD/<event_type>-<YYYYMMDD>-<uuid>.ndjson

Files are opened in append mode; multiple events from the same process,
date, and event-type share a single file. The UUID suffix is stable for
the lifetime of the process, keeping one file per (process × type × day).
"""

import json
import logging
import os
import threading
import uuid
from datetime import datetime, timezone
from typing import Any, Dict

try:
    from loguru import logger
except ImportError:
    logger = logging.getLogger(__name__)


def _serialize(obj: Any) -> Any:
    """JSON-serialise types that json.dumps doesn't handle natively."""
    if isinstance(obj, datetime):
        return obj.astimezone(timezone.utc).isoformat()
    raise TypeError(f"Object of type {type(obj).__name__} is not JSON serialisable")


class LocalBackend:
    """Appends audit events to local NDJSON files."""

    def __init__(self, path: str) -> None:
        self._root = path
        self._file_uuid = uuid.uuid4().hex[:12]
        self._lock = threading.Lock()

    def write(self, event: Dict[str, Any]) -> None:
        """Append *event* as a JSON line. Never raises; logs errors."""
        with self._lock:
            self._write_locked(event)

    def _write_locked(self, event: Dict[str, Any]) -> None:
        try:
            ts: datetime = event.get("ts") or datetime.now(timezone.utc)
            if isinstance(ts, str):
                ts = datetime.fromisoformat(ts.replace("Z", "+00:00"))
            utc = ts.astimezone(timezone.utc)

            event_type: str = str(event.get("type", "audit"))
            date_dir = os.path.join(
                self._root,
                utc.strftime("%Y"),
                utc.strftime("%m"),
                utc.strftime("%d"),
            )
            os.makedirs(date_dir, exist_ok=True)

            filename = f"{event_type}-{utc.strftime('%Y%m%d')}-{self._file_uuid}.ndjson"
            filepath = os.path.join(date_dir, filename)

            line = json.dumps(event, default=_serialize) + "\n"
            with open(filepath, "a", encoding="utf-8") as fh:
                fh.write(line)
        except Exception as exc:  # noqa: BLE001
            logger.warning(f"[audit/local] Failed to write audit event: {exc}")
